xenserver/task_lease.py

```diff
@@ -0,0 +1,51 @@
+"""
+This module (ab)uses the database session store to provide task leases that
+can be used to prevent a buildup of slow tasks.
+"""
+
+from __future__ import absolute_import
+
+from datetime import timedelta
+
+from celery.utils.log import get_task_logger
+from django.conf import settings
+from django.contrib.sessions.models import Session
+from django.utils import timezone
+
+logger = get_task_logger(__name__)
+
+
+def _key(task, args):
+    return '::'.join(['lease', task, args])
+
+
+def acquire(task, args, ttl=None):
+    """
+    Attempt to acquire a task lease for the given task. This should be done by
+    the thing that queues the task for execution. Returns `True` is the lease
+    has been acquired, `False` if an existing active lease is found.
+    """
+    key = _key(task, args)
+    if Session.objects.filter(
```

Comment:
`get_or_create` doesn't let me check the `expire_date` in the query, but it does require me to pass it as a default for the creation step. While this makes the creation of a new object easier (it happens as part of `get_or_create`), it makes handling an existing but expired object harder because I'd have to check and update `expire_date` as appropriate. It's about the same amount of code as doing it this way, but the logic for checking if we properly got the lease is split up and less readable.